src/hotspot/share/gc/shared/referenceProcessor.cpp

```diff
@@ -1068,75 +1068,35 @@ void ReferenceProcessor::preclean_discovered_references(BoolObjectClosure* is_al
                                                         EnqueueDiscoveredFieldClosure* enqueue,
                                                         YieldClosure* yield,
                                                         GCTimer* gc_timer) {
-  // These lists can be handled here in any order and, indeed, concurrently.
+  Ticks preclean_start = Ticks::now();
 
-  // Soft references
-  {
-    GCTraceTime(Debug, gc, ref) tm("Preclean SoftReferences", gc_timer);
-    log_reflist("SoftRef before: ", _discoveredSoftRefs, _max_num_queues);
-    for (uint i = 0; i < _max_num_queues; i++) {
-      if (yield->should_return()) {
-        return;
-      }
-      if (preclean_discovered_reflist(_discoveredSoftRefs[i], is_alive,
-                                      enqueue, yield)) {
-        log_reflist("SoftRef abort: ", _discoveredSoftRefs, _max_num_queues);
-        return;
-      }
-    }
-    log_reflist("SoftRef after: ", _discoveredSoftRefs, _max_num_queues);
-  }
+  ReferenceType ref_type_arr[4] = { REF_SOFT, REF_WEAK, REF_FINAL, REF_PHANTOM };
+  size_t ref_count_arr[4] = {};
 
-  // Weak references
-  {
-    GCTraceTime(Debug, gc, ref) tm("Preclean WeakReferences", gc_timer);
-    log_reflist("WeakRef before: ", _discoveredWeakRefs, _max_num_queues);
-    for (uint i = 0; i < _max_num_queues; i++) {
-      if (yield->should_return()) {
-        return;
-      }
-      if (preclean_discovered_reflist(_discoveredWeakRefs[i], is_alive,
-                                      enqueue, yield)) {
-        log_reflist("WeakRef abort: ", _discoveredWeakRefs, _max_num_queues);
-        return;
-      }
+  if (discovery_is_mt()) {
```

Comment:
The old form of these loops called `yield->should_return()` for each queue.  That's been lost in this rewrite, and I'm not sure that's correct.

Reply:
It's still correct because `preclean_discovered_reflist` checkes for if-aborted while iterating the list. I could have written sth like the following if you think prompt abort is important.

```
  bool is_aborted = preclean_discovered_reflist(...);
  if (is_aborted) {
    return;
  }
```

Reply:
It's hard to know for sure, but the description of YieldClosure makes me think that's not really the intended usage.  OTOH, this seems to be the only use of YieldClosure; maybe there were others in CMS?  I would prefer the `yield->should_return()` checks be retained.  Also, `preclean_discovered_reflist` might no longer need to return bool.

Reply:
Restored `yield->should_return()`.